sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -620,6 +628,106 @@ def _next_id(self):
     return str(self._last_id)
 
 
+class CachingMaterializingStateHandler(object):
+  """ A State handler which retrieves and caches state. """
+
+  def __init__(self, global_state_cache, underlying_state):
+    self._underlying = underlying_state
+    self._state_cache = global_state_cache
+    self._context = threading.local()
+
+  @contextlib.contextmanager
+  def process_instruction_id(self, bundle_id, cache_tokens):
+    if getattr(self._context, 'cache_token', None) is not None:
+      raise RuntimeError(
+          'Cache tokens already set to %s' % self._context.cache_token)
+    # TODO Also handle cache tokens for side input, if present
```

Comment:
Create JIRA for it?

Reply:
Will do!

Reply:
https://issues.apache.org/jira/browse/BEAM-8297

Reply:
https://issues.apache.org/jira/browse/BEAM-8298